File: book8-cracking-the-coding-interview/preface-vi-big-o/Example11Factorial.java
public class Example11Factorial {

    /**
     * Factorial recursive.
     * Time: O(n)
     * Memory: O(n)
     * 
     * @param n number
     * @return n!
     */
    int factorialRecursive(int n) {
        if (n < 0) {
            return -1;
        } else if (n == 0) {
            return 1;
        } else {
            return n * factorialRecursive(n - 1);
        }
    }

    /**
     * Factorial iterative.
     * Time: O(n)
     * Memory: O(1)
     * 
     * @param n number
     * @return n!
     */
    int factorialIterative(int n) {
        int result = n--;
        while (n >= 1) {
            result *= n--;
        }
        return result;
    }

    void run() {
        long start = System.nanoTime();
        int result = factorialRecursive(15);
        long end = System.nanoTime();
        long delta = end - start;
        System.out.println("FactorialRecursive(10): " + result + " " + delta + "ns");
        start = System.nanoTime();
        result = factorialIterative(15);
        end = System.nanoTime();
        delta = end - start;
        System.out.println("FactorialIterative(10): " + result + " " + delta + "ns");
    }

    public static void main(String[] args) {
        new Example11Factorial().run();
    }
}

File: book8-cracking-the-coding-interview/preface-vi-big-o/Example11Factorial.go
package main

import (
	"fmt"
	"time"
)

func FactorialRecursive(n int) int {
	if n < 0 {
		return -1
	} else if n <= 1 {
		return 1
	} else {
		return n * FactorialRecursive(n-1)
	}
}

func FactorialIterative(n int) int {
	result := 1
	for ; n >= 1; n-- {
		result = result * n
	}
	return result
}

func main() {
	start := time.Now()
	res := FactorialRecursive(10)
	duration := time.Since(start)
	fmt.Printf("FactorialRecursive(10)=%v %vns\n", res, duration.Nanoseconds())
	start = time.Now()
	res = FactorialIterative(10)
	duration = time.Since(start)
	fmt.Printf("FactorialIterative(10)=%v %vns\n", res, duration.Nanoseconds())
}
